WIP/draft: enarx-wasmldr updates & refactoring

wgwoods opened the pull request:
This is a work-in-progress patch series to get enarx-wasmldr working with current wasmtime (0.29). Since `virtfs` was removed from upstream `wasmtime_wasi`, we also have to rethink how we bundle/include configuration and assets.

Current status: `enarx-wasmldr` compiles against wasmtime 0.29 and all its tests pass. As before, It can load modules from the commandline or from a file descriptor - the FD number can be chosen using the `--module-on-fd FDNUM` argument.

WASI stdin/stdout/stderr are inherited from the parent process. (CLI args / config options to control that will probably get added later.) WASI environment variables are *not* inherited from the parent process, but you can set them manually with the `-e`/`--env` flag. CLI args after `--` get passed into the WASI environment, as you'd expect.

The `--wasm-features` flag is parsed correctly by the CLI but doesn't do anything at the moment. But "module-linking" and "multi-memory" are turned on, so you can run modules that use module-linking.

There's no configuration format or file embedding, yet. I'm hoping this is just enough to let us work on a front-end tool / `enarx run` command that wraps/embeds wasmldr and can actually run WebAssembly modules inside a proper keep.

Comments? Suggestions? Any vital functionality that we're missing here?

wgwoods committed:
add @wgwoods to CODEOWNERS

Signed-off-by: Will Woods <will@congru.us>

npmccallum commented:
@wgwoods Without any further review there are three things you need to fix:
1. Make `cargo clippy` succeed.
2. Generate your readme with `cargo readme > README.md`
3. Sign-off on all our commits (i.e. `git commit -s`)

wgwoods committed:
Bump MSRV in test.yml to 1.53.0

currently, test.yml has rust 1.46.0. But...

wasmtime 0.29 (and a few of its deps) require cargo's "resolver=2",
which raises our Minimum Supported Rust Version of 1.51.0.

At least one other dep (posish) needs 1.53.0, so that bumps our MSRV
again.

Signed-off-by: Will Woods <will@congru.us>

wgwoods committed:
remove .cargo/config

Choosing the appropriate target is going to be Someone Else's Problem,
and I want to be able to build/test things locally (esp. now that
wasmtime builds on aarch64-apple-darwin, hooray)

Signed-off-by: Will Woods <will@congru.us>

wgwoods committed:
Rework for wasmtime 0.29 and WASI changes

A lot has changed in upstream wasmtime / WASI, so wasmldr needs
significant reworking to keep up.

The biggest change upstream is that virtfs has been dropped from
wasi_common. So we're dropping the bundle format and virtual filesystem
view, for now, and we'll reimplement that functionality piece by piece,
using things that are a little better aligned with current directions in
WASI and WebAssembly.

With this commit, enarx-wasmldr at least compiles and passes the remaining
tests, and can successfully run some random wasm modules I had laying
around. So this seems like a good start.

Detailed changes:

* bumped deps to wasmtime 0.29 and wasmparser 0.80
* removed bundle stuff from build.rs
* removed config.rs and virtfs.rs
* removed deps on serde, tar, tempfile from Cargo.toml
* remove populate_virtfs() etc. from workload.rs
  * NOTE: this means we can't load modules on FD3 right now (that will
    be restored in a future commit)
* moved CLI parsing into cli.rs, since this might get reused elsewhere
* switched to structopt for CLI parsing, as we do elsewhere
* use anyhow for error handling (like wasmtime does)
* tweak trait bounds on workload::run() to work better with args/envs
  from structopt
* enable module-linking and multi-memory in instantiated wasmtime engine

Signed-off-by: Will Woods <will@congru.us>

wgwoods committed:
Add --module-from-fd FD and get stdio working

This is definitely a work in progress... but hey, progress:

* Added '--module-from-fd NUM' arg, which lets you specify which FD to
  read the module from
* Brought back config.rs and the DeployConfig stuff
* Fixed up the header comment in main.rs to reflect current reality

Caveats / things to fix Real Soon Now:

* stdin/stdout/stderr are currently hardcoded to Inherit
* engine features are hardcoded to include module-linking &
  multi-memory
* workload::Error has a half-assed implementation of Display

wgwoods committed:
CLI improvements

* Dropped TrailingVarArg, so '--' is now required to separate wasmldr
  args from args to be passed into WasiCtx
* Fleshed out --wasm-features parsing
* Added lazy_static to generate help text for --wasm-features
* --module-on-fd refuses to use any fd < 3

wgwoods committed:
code cleanups in workload.rs

* silence warnings about unused error variants (for now..)
* clean up comments so I'm using TODO everywhere
* drop unused import

getting ready for some refactoring here. wheeeee!

wgwoods commented:
This has been obsoleted by other work, so I'm closing it.
